hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+use std::fmt;
+
+use cranelift_entity::{entity_impl, PrimaryMap};
+use winter_math::FieldElement;
+
+use super::{write::DisplayIndent, *};
+
+/// Represents Miden Assembly (MASM) directly in the IR
+///
+/// Each block of inline assembly executes in its own pseudo-isolated environment,
+/// i.e. other than arguments provided to the inline assembly, and values introduced
+/// within the inline assembly, it is not permitted to access anything else on the
+/// operand stack
+#[derive(Debug, Clone)]
+pub struct InlineAsm {
+    pub op: Opcode,
+    /// Arguments on which the inline assembly can operate
+    ///
+    /// The operand stack will be set up such that the given arguments
+    /// will appear in LIFO order, i.e. the first argument will be on top
+    /// of the stack, and so on.
+    ///
+    /// The inline assembly will be validated so that all other values on
+    /// the operand stack below the given arguments will remain on the stack
+    /// when the inline assembly finishes executing.
```

Comment:
I just noticed that the behavior I describe is indeed allowed, but it's only described in the comments for the builder functions. 

I think it would be a good idea to flesh out these kinds of assumptions either in the comments of `InlineAsm`, or in separate documentation. People will be looking for those descriptions there, not in the builder function comments.

Reply:
Inline assembly blocks can support zero or more results, much like function calls. The output of an inline assembly block is given by the result type, which in the text format starts after the trailing `:`. For now though, that's specified using a single `Type` via the builders, so representing zero, one, or more results requires using an appropriate `Type`, e.g. `Type::Unit`, `Type::Felt`, or `Type::Array(Box::new(Type::Felt), n)` for 0, 1, or n-ary results.

> I think it would be a good idea to flesh out these kinds of assumptions either in the comments of InlineAsm, or in separate documentation. People will be looking for those descriptions there, not in the builder function comments.

Could you clarify what assumptions you are referring to here? `InlineAsm` doesn't impose any restrictions on the number or type of results on its own, and what restrictions exist are imposed by the builders (either the `inline_asm` of `InstBuilder`, or by `MasmBuilder`), and their docs reflect those restrictions (unless I missed one somewhere). In general, it is expected that users of the IR (i.e. compiler frontends) are going to work through the builders, and not need to be aware of `InlineAsm` unless they are doing something low-level with the IR, I would be very surprised if the first thing someone looked at was that struct, vs the docs associated with the builders.

I'm hesitant to go too far down the path of separate high-level documentation at this stage, there is still a lot of instability in the IR that remains to be ironed out, and high-level docs are particularly prone to falling out of date, and can be difficult to maintain under a stream of constant change. I'm as keen to get those kind of docs in place as you are, but I want to focus on stabilizing the foundation first, then start building stuff on it.

That said, if you find that there are aspects of the implementation that are undocumented, I would like to make sure that is addressed. I will expand on the `InlineAsm` docs a bit to provide at least a little more overview on it's structure and usage.

Reply:
I'm not talking about fully fledged documentation at this stage. What I'm talking about at this stage is much more low-level, and could be as simple as an unsorted list of design decisions that have been made along the way, and which could eventually be collated into actual user-facing documentation.

The alternative is to remember everything in our heads, and in my experience that always leads to problems later on.